### PR TITLE
Change the default location for Enso projects

### LINKED ISSUE #10240: Change the default location for Enso project on Mac and Windows
Should be `My Documents/enso-projects` (or `Documents/enso-projects` on Mac.

However, if the user already has the current structure we should move it to the new location.

If starting enso on a complete pristine machine (i.e. no folder) it should be created and not crash!

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #10240 

Changelog:
- add: `desktop-environment` Java module to detect user environment configuration
- add: `ProjectsMigration` module containing the migration logic of the enso projects directory
- update: updated and cleaned up unused settings from the storage config
- add: `desktopEnvironment` TS module to detect user environment configuration in the `project-manager-shim`
- update: `project-manager-shim` with the new user projects directory

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

**Review by somebody1234 — APPROVED** [review]
approving changes to the PM shim used by GUI2's dev server

File: build.sbt [hunk 1]
```diff
@@ -2791,6 +2792,17 @@ lazy val `benchmarks-common` =
     )
     .dependsOn(`polyglot-api`)
 
+lazy val `desktop-environment` =
```
Comment (JaroslavTulach) [inline build.sbt:2795]: No 3rd party dependencies. I like such _lightweight_ modules.

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/Platform.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  private static final String MAC = "mac";
+  private static final String WINDOWS = "windows";
+
+  private Platform() {}
+
+  public static Directories getDirectories() {
+    return DirectoriesFactory.getInstance();
+  }
+
+  public static String getOsName() {
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/Platform.java:12]: If you have to define the `getOsName()` method at all...

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/Platform.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    return System.getProperty(OS_NAME);
+  }
+
+  public static boolean isLinux() {
+    return System.getProperty(OS_NAME).toLowerCase().contains(LINUX);
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/Platform.java]: ...then you should use the `getOsName()` method for example here.

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/Platform.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package org.enso.desktopenvironment;
+
+import org.enso.desktopenvironment.directories.Directories;
+import org.enso.desktopenvironment.directories.DirectoriesFactory;
+
+public final class Platform {
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/Platform.java:3]:
```java
enum Platform {
  LINUX, MAC, WINDOWS
}
```
might be better choice from an API perspective.

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesException.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package org.enso.desktopenvironment.directories;
+
+import java.io.IOException;
+
+/** Indicates an issue when accessing user directories. */
+public final class DirectoriesException extends IOException {
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesException.java]: Any benefit of having own `IOException` subclass?

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesException.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.enso.desktopenvironment.directories;
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesException.java]: I'd flatten the directories and put all API into `org.enso.desktopenvironment` package.

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesFactory.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package org.enso.desktopenvironment.directories;
+
+import org.enso.desktopenvironment.Platform;
+
+public final class DirectoriesFactory {
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesFactory.java]: Please simplify. No _factory_, unless really needed. In my opinion a getter in `Platform` is a good way to obtain `Directories` instance. No need to expose _factory_ to public.

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/WindowsDirectories.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+   *
+   * <p>Method uses the registry query that may not work on Windows XP versions and below.
+   *
+   * @return the 'My Documents' user directory path.
+   * @throws DirectoriesException when fails to detect the user documents directory.
+   */
+  @Override
+  public Path getDocuments() throws DirectoriesException {
+    try {
+      var process = new ProcessBuilder(PROCESS_REG_QUERY).start();
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/WindowsDirectories.java]: Avoiding a process execution would be nicer, but not even [stack over flow has a suggestion](https://stackoverflow.com/questions/62289/read-write-to-windows-registry-using-java) how to do that nicely.

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/LinuxDirectories.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  public Path getDocuments() {
+    try {
+      return getXdgDocuments();
+    } catch (IOException | InterruptedException e) {
+      return getUserHome().resolve("enso");
+    }
+  }
+
+  private Path getXdgDocuments() throws IOException, InterruptedException {
+    var process = new ProcessBuilder(PROCESS_XDG_DOCUMENTS).start();
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/LinuxDirectories.java]: Thanks for using `xdg-user-dir` I am looking forward for Enso to put the projects into the right directory on my Linux box!

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/Directories.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package org.enso.desktopenvironment.directories;
+
+import java.nio.file.Path;
+
+public interface Directories {
```
Comment (JaroslavTulach) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/Directories.java]: Do you want other code outside of this module to implement this interface? If not, then prevent it:
- (in old Java days) I'd recommend `abstract class` with _package private_ constructor
- now there is also an option to defined `sealed` interface

File: lib/scala/project-manager/src/main/java/org/enso/projectmanager/event/package-info.java
Comment (JaroslavTulach) [inline lib/scala/project-manager/src/main/java/org/enso/projectmanager/event/package-info.java:1]: Why are you deleting this file? Are the reasons for its existence obsolete?

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/configuration.scala [hunk 11]
```diff
@@ -41,21 +44,37 @@ object configuration {
...
 
+  /** A configuration object for properties of project storage.
+    *
+    * @param projectsRoot overrides user projects root directory
+    * @param projectsDirectory a user projects directory
+    * @param metadata a metadata storage config
+    */
+  case class StorageConfig(
+    projectsRoot: Option[File],
+    projectsDirectory: String,
```
Comment (JaroslavTulach) [inline lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/configuration.scala:64]: `File` for `projectRoot` and `String` for a directory? Why not use `File` for both?

**Review by JaroslavTulach — APPROVED** [review]

File: lib/scala/project-manager/src/main/java/org/enso/projectmanager/event/package-info.java
Comment (4e6) [reply lib/scala/project-manager/src/main/java/org/enso/projectmanager/event/package-info.java:1]: It was added in #4098 as a dummy Java file to be able to open the project manager in IVG.

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesException.java (hunk 5 above)
Comment (4e6) [reply lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/DirectoriesException.java]: Removed

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/Directories.java (hunk 10 above)
Comment (4e6) [reply lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/directories/Directories.java]: good old Java days :cowboy_hat_face:

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/configuration.scala (hunk 11 above)
Comment (4e6) [reply lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/configuration.scala:64]: File values represent paths, and string values represent names

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/LinuxDirectories.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+   * available and falls back to {@code $HOME/enso}.
+   *
+   * @return the path to the user documents directory.
+   */
+  @Override
+  public Path getDocuments() {
+    try {
+      return getXdgDocuments();
+    } catch (IOException | InterruptedException e) {
+      return getUserHome().resolve("enso");
```
Comment (hubertp) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/LinuxDirectories.java:25]: Log a warning about a fallback?

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/MacOsDirectories.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+   *
+   * @return the path to the user documents directory.
+   * @throws IOException when unable to resolve the real documents path.
+   */
+  @Override
+  public Path getDocuments() throws IOException {
+    try {
+      return getUserHome().resolve(DOCUMENTS).toRealPath();
+    } catch (IOException e) {
+      throw new IOException("Failed to resolve real MacOs documents path", e);
```
Comment (hubertp) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/MacOsDirectories.java:24]: Why no fallback to whatever we have now?

File: lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/WindowsDirectories.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+      var stdoutString =
+          new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
+      var stdoutParts = stdoutString.split("\\s\\s+");
+      if (stdoutParts.length < 5) {
+        throw new IOException("Invalid Windows registry query output: '" + stdoutString + "'");
+      }
+
+      return Path.of(stdoutParts[4].trim());
+    } catch (IOException e) {
+      throw new IOException("Failed to run Windows registry query", e);
```
Comment (hubertp) [inline lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/WindowsDirectories.java:42]: No fallback?

**Pull request merged**